test/test-sock-send-recv.c

```diff
@@ -0,0 +1,162 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+#include "uv.h"
+#include "uvwasi.h"
+#include "wasi_serdes.h"
+#include <stdio.h>
+
+#define PREOPEN_SOCK 3
+#define INVALID_SOCK 42
+#define DEFAULT_BACKLOG 5
+
+int delayedThreadTime =  5000;
+int immedateThreadTime =  0;
+
+#define CONNECT_ADDRESS "127.0.0.1"
+#define TEST_PORT_1 10500
+
+void on_uv_close(uv_handle_t* handle) {
+  free(handle);
+}
+
+static void alloc_cb(uv_handle_t* handle, size_t suggested_size, uv_buf_t* buf) {
+  buf->base = malloc(suggested_size);
+  buf->len = suggested_size;
+}
+
+void echo_read(uv_stream_t *stream, ssize_t nread, const uv_buf_t *buf) {
+  if (nread > 0) {
+    uv_buf_t send_buf;
+    uv_read_stop(stream);
+    send_buf.base = buf->base;
+    send_buf.len = nread;
+    uv_try_write(stream, &send_buf, 1);
+  }
+  uv_close((uv_handle_t *) stream, on_uv_close);
+  free(buf->base);
+}
+
+void on_client_connect(uv_connect_t * req, int status) {
+  if (status < 0) {
+      fprintf(stderr, "New connection error %s\n", uv_strerror(status));
+      // error!
+      return;
+  }
+  uv_read_start((uv_stream_t*) req->handle, alloc_cb, echo_read);
+  free(req);
+}
+
+void client_connection_echo_thread(void* time) {
+  int r = 0;
+  uv_loop_t *loop = malloc(sizeof(uv_loop_t));
```

Comment:
I think you could reduce the number of `malloc()` and `free()` calls by creating static variables for the test.